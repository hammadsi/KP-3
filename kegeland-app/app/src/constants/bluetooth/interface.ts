import {DeviceStackParamList} from '~routes/interface';

export type BluetoothCharacteristic = string;
export type BluetoothService = {
  uuid: string;
  characteristics: BluetoothCharacteristic[];
};

export type BluetoothProfile = {
  navScreen: keyof DeviceStackParamList;
  services: string[];
  batteryService: {
    service: BluetoothService;
    fn: (bytes: number[]) => number;
  };
};

export type PeripheralNotification = {
  value: number[];
  peripheral: string;
  characteristic: string;
  service: string;
};
