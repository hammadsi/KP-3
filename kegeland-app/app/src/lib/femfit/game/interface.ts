import Matter from 'matter-js';
import {GameEngineUpdateEventOptionType} from 'react-native-game-engine';

export type Position = {
  x: number;
  y: number;
};

export type EntityBase = {
  body: Matter.Body;
};

export type ExerciseScheme = {
  name: string;
  description: string;
  repetitions: number;
  data: [number, number][];
};

export type GameEntities = {
  [key: string]: EntityBase & {[key: string]: any};
} & {
  physics: {
    engine: Matter.Engine;
    world: Matter.World;
  };
  exercise: ExerciseScheme & {currRep: number; currStep: number};
};

export type IGameEngineSystem = (
  entities: GameEntities,
  update: GameEngineUpdateEventOptionType,
) => GameEntities;
