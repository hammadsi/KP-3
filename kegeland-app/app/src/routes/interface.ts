import {
  CompositeScreenProps,
  NavigatorScreenParams,
} from '@react-navigation/native';
import type {BottomTabScreenProps} from '@react-navigation/bottom-tabs';
import type {StackScreenProps} from '@react-navigation/stack';

/**
 * Navigation interface for auth routes
 */
export type AuthStackParamList = {
  Login: undefined;
  Register: undefined;
  'Forgot password': undefined;
};

export type DeviceStackParamList = {
  Devices: undefined;
  Femfit: undefined;
};

/**
 * Navigation interface for settings routes
 */
export type SettingsStackParamList = {
  Settings: undefined;
  'Connect device': undefined;
};

/**
 * Base navigation interface
 */
export type RootTabParamList = {
  AuthStack: NavigatorScreenParams<AuthStackParamList> | undefined;
  DeviceStack: NavigatorScreenParams<DeviceStackParamList> | undefined;
  SettingsStack: NavigatorScreenParams<SettingsStackParamList> | undefined;
};

/**
 * Interface for screens on root-level
 */
export type RootScreenProps<
  RouteName extends keyof RootTabParamList = keyof RootTabParamList,
> = BottomTabScreenProps<RootTabParamList, RouteName>;

/**
 * Interface for screens under the auth-stack
 */
export type AuthScreenProps<
  RouteName extends keyof AuthStackParamList = keyof AuthStackParamList,
> = CompositeScreenProps<
  StackScreenProps<AuthStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

export type DeviceScreenProps<
  RouteName extends keyof DeviceStackParamList = keyof DeviceStackParamList,
> = CompositeScreenProps<
  StackScreenProps<DeviceStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;

/**
 * Interface for screens under the settings-stack
 */
export type SettingsScreenProps<
  RouteName extends keyof SettingsStackParamList = keyof SettingsStackParamList,
> = CompositeScreenProps<
  StackScreenProps<SettingsStackParamList, RouteName>,
  BottomTabScreenProps<RootTabParamList>
>;
