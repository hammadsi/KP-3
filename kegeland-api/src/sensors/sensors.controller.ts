import {
  Controller,
  Get,
  Post,
  Put,
  Delete,
  Body,
  Param,
  UseGuards,
} from '@nestjs/common';
import { ApiBearerAuth, ApiTags } from '@nestjs/swagger';

import { Role } from '../roles/enums/role.enum';
import { Roles } from '../roles/roles.decorator';
import { RolesGuard } from '../roles/roles.guard';
import { FirebaseAuthGuard } from '../firebase/firebase-auth.guard';

import { UpdateSensorDto } from './dto/update-sensor.dto';
import { CreateSensorDto } from './dto/create-sensor.dto';
import { SensorsService } from './sensors.service';

@ApiTags('Sensors')
@Controller('sensors')
@ApiBearerAuth('access-token')
@UseGuards(FirebaseAuthGuard, RolesGuard)
@Roles(Role.PATIENT, Role.PHYSICIAN)
export class SensorsController {
  constructor(private readonly sensorsService: SensorsService) {}

  /**
   * Endpoint for fetching specific sensor
   * @param id of sensor
   * @returns a Sensor object
   */
  @Get(':id')
  async findOne(@Param('id') id: string) {
    return this.sensorsService.findOne(id);
  }

  /**
   * Endpoint for fetching all sensors in the system
   * @returns list of all Sensors in the database
   */
  @Get()
  async findAll() {
    return this.sensorsService.findAll();
  }

  /**
   * Endpoint for updating existing sensor
   * @param id of sensor
   * @param updateSensorDto fields and data to be updated passed as body in request
   * @returns updated Sensor object
   */
  @Put(':id')
  async update(
    @Param('id') id: string,
    @Body() updateSensorDto: UpdateSensorDto,
  ) {
    return this.sensorsService.update(id, updateSensorDto);
  }

  /**
   * Endpoint for creating new sensor
   * @param createSensorDto data for new sensor passed as body in request
   * @returns the created Sensor object
   */
  @Post()
  async create(@Body() createSensorDto: CreateSensorDto) {
    return this.sensorsService.create(createSensorDto);
  }

  /**
   * Endpoint for deleting sensor
   * @param id of sensor to be deleted
   * @returns Promise<void>
   */
  @Delete(':id')
  async delete(@Param('id') id: string) {
    return this.sensorsService.delete(id);
  }
}
