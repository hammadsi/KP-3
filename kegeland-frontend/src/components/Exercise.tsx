import {
  Box,
  Text,
  WrapItem,
  Divider,
  LinkBox,
  LinkOverlay,
} from '@chakra-ui/react';
import { useParams } from 'react-router-dom';

interface ExerciseProps {
  date: string;
  id: string;
  sensor: string;
}

const Exercise: React.FC<ExerciseProps> = ({ date, id, sensor }) => {
  const { patientId } = useParams();

  return (
    <WrapItem>
      <LinkBox
        w="500px"
        p={['12', '12']}
        bg="gray.100"
        borderRadius={20}
        pb="4">
        <LinkOverlay
          href={'/patient/' + patientId + '/exercise/' + id}
          display="flex-center"
          justifyContent="space-between">
          <Box>{date}</Box>
          <Box>{sensor}</Box>
        </LinkOverlay>
        <Divider orientation="horizontal" borderColor={'blue.900'} />
        <Text fontSize="lg" mb="4"></Text>
      </LinkBox>
    </WrapItem>
  );
};

export default Exercise;
