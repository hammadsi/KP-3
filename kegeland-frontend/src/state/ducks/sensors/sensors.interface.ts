export enum SensorType {
  FEMFIT = 'femfit',
  EMPATICA = 'empatica',
  IMU = 'imu',
}

export type Sensor = {
  id: string;
  name: SensorType;
  labels: string[];
};

export interface SensorState {
  loading: boolean;
  data: {
    [key in SensorType]?: Sensor;
  };
  error: string | undefined;
}
