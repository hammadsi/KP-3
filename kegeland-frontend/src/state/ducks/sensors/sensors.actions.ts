import { createAsyncThunk } from '@reduxjs/toolkit';

import { apiCaller } from '../../../utils/apiCaller';

import { Sensor, SensorType } from './sensors.interface';

export const fetchSensor = createAsyncThunk(
  'sensors/fetchSensor',
  async (sensorType: SensorType) =>
    apiCaller<Sensor>({
      url: `sensors/${sensorType}`,
      method: 'GET',
    }),
);
