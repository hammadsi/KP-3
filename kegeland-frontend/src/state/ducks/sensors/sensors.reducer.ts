import { createSlice } from '@reduxjs/toolkit';

import {
  isPendingAction,
  isFulfilledAction,
  isRejectedAction,
} from '../../../utils/thunk.utils';

import { fetchSensor } from './sensors.actions';
import { SensorState } from './sensors.interface';

export const initialState: SensorState = {
  loading: false,
  data: {},
  error: undefined,
};

const sensorSlice = createSlice({
  name: 'sensors',
  initialState,
  reducers: {
    clearSensorsState: (state) => {
      state.loading = false;
      state.data = {};
      state.error = undefined;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchSensor.fulfilled, (state, { payload }) => {
        state.data[payload.name] = payload;
      })
      .addMatcher(
        (action) => isPendingAction(action, 'sensors'),
        (state) => {
          state.loading = true;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isFulfilledAction(action, 'sensors'),
        (state) => {
          state.loading = false;
          state.error = undefined;
        },
      )
      .addMatcher(
        (action) => isRejectedAction(action, 'sensors'),
        (state, { error }) => {
          state.loading = false;
          state.error = error.message;
        },
      );
  },
});

export const { clearSensorsState } = sensorSlice.actions;

export const sensorReducer = sensorSlice.reducer;
