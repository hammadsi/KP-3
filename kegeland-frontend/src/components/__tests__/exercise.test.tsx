import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import { SensorType } from '../../state/ducks/sensors/sensors.interface';
import Exercise from '../Exercise';

describe('Test excercise', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <Exercise date="20.02.2022" id="id" sensor={SensorType.FEMFIT} />{' '}
        </MemoryRouter>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
