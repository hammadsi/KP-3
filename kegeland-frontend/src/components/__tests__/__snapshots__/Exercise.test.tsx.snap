// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test excercise renders correctly 1`] = `
Array [
  <li
    className="chakra-wrap__listitem css-1yp4ln"
  >
    <div
      className="chakra-linkbox css-rd9efr"
    >
      <a
        className="chakra-linkbox__overlay css-zpisrh"
        href="/patient/undefined/exercise/id"
      >
        <div
          className="css-0"
        >
          20.02.2022
        </div>
        <div
          className="css-0"
        >
          femfit
        </div>
      </a>
      <hr
        aria-orientation="horizontal"
        className="chakra-divider css-3f0eww"
      />
      <p
        className="chakra-text css-x4lmqn"
      />
    </div>
  </li>,
  " ",
]
`;
